nengo_gui/static/components/spa_similarity.js

```diff
@@ -0,0 +1,202 @@
+};
+
+Nengo.SpaSimilarity.prototype.update_legend = function(new_labels){
+    // Should figure out how to mix recs and text into one
+
+    var self = this;
+    this.pointer_labels = this.pointer_labels.concat(new_labels);
+
+    // expand the svg
+    this.legend_svg.attr("height", 20*this.pointer_labels.length);
+
+    // Data join
+    var recs = this.legend_svg.selectAll("rect").data(this.pointer_labels);
+    var texts = this.legend_svg.selectAll("text").data(this.pointer_labels);
+    // nothing to update
+    // enter to append remaining lines
+    recs.enter()
+        .append("rect")
+        .attr("x", 0)
+        .attr("y", function(d, i){ return i *  20;})
+        .attr("width", 10)
+        .attr("height", 10)
+        .style("fill", this.color_func);
+
+    texts.enter()
+          .append("text")
+          .attr("x", 15)
+          .attr("y", function(d, i){ return i *  20 + 9;})
+          .html(function(d, i) {
+                return self.pointer_labels[i];
+           });
+
+};
+
+/* there a three types of messages that can be recieved:
+    - a legend needs to be updated
+    - the data has been updated
+    - show_pairs has been toggled
+    this calls the method associated to handling the type of message
+*/
+Nengo.SpaSimilarity.prototype.on_message = function(event) {
+    var data = JSON.parse(event.data);
+    var func_name = data.shift();
+    this[func_name](data);
+};
+
+/**
+ * Redraw the lines and axis due to changed data
+ */
+Nengo.SpaSimilarity.prototype.update = function() {
+    /** let the data store clear out old values */
+    this.data_store.update();
+
+    /** determine visible range from the Nengo.SimControl */
+    var t1 = this.sim.time_slider.first_shown_time;
+    var t2 = t1 + this.sim.time_slider.shown_time;
+
+    this.axes2d.set_time_range(t1, t2);
+
+    /** update the lines */
+    var self = this;
+    var shown_data = this.data_store.get_shown_data();
+    // Data join
+    this.path = this.axes2d.svg.selectAll(".line").data(shown_data);
+    // update
+    this.path.attr('d', self.line);
+    // enter to append remaining lines
+    this.path.enter()
+             .append('path')
+             .attr('class', 'line')
+             .style('stroke', this.color_func)
+             .attr('d', self.line);
+    // remove any lines that aren't needed anymore
+    this.path.exit().remove();
+
+    /* update the legend text */
+    if(this.legend_svg && shown_data[0].length !== 0){
+        // get the most recent similarity
+         var latest_simi = [];
+        for(var i = 0; i < shown_data.length; i++){
+            latest_simi.push(shown_data[i][shown_data[i].length - 1]);
+        }
+
+        // update the text in the legend
+        var texts = this.legend_svg.selectAll("text").data(this.pointer_labels);
+
+        texts.attr("x", 15)
+              .attr("y", function(d, i){ return i *  20 + 9;})
+              .html(function(d, i) {
+                    var sign = "&nbsp;&nbsp;&nbsp;";
```

Comment:
How difficult would it be to use an invisible + instead of non-breaking spaces? I'm still experiencing some jitter.